engine/crates/runtime/src/hooks.rs

```diff
@@ -115,6 +123,191 @@ pub trait SubgraphHooks<Context>: Send + Sync + 'static {
     ) -> impl Future<Output = Result<HeaderMap, PartialGraphqlError>> + Send;
 }
 
+#[derive(Debug, Clone, Copy)]
+pub struct ResponseInfo {
+    pub connection_time: u64,
+    pub response_time: u64,
+    pub status_code: u16,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum ResponseKind {
+    SerializationError,
+    HookError,
+    RequestError,
+    RateLimited,
+    Responsed(ResponseInfo),
+}
+
+impl ResponseInfo {
+    pub fn builder() -> ResponseInfoBuilder {
+        ResponseInfoBuilder {
+            start: SystemTime::now(),
+            connection_time: None,
+            response_time: None,
+        }
+    }
+}
+
+pub struct ResponseInfoBuilder {
+    start: SystemTime,
```

Comment:
`Instant` would be more appropriate for a duration, it has the convenient function `elapsed()` and it's guaranteed to not be negative